allow setting single-az with multiple nodes

cabrinha opened the pull request:
## what
* I want to deploy multiple nodes into a single AZ.

## why
* I run a high throughput, low-latency service in a single AZ. I'd like more elasticache nodes in that AZ.

Gowiem commented:
/test all

Gowiem commented on variables.tf:
```diff
@@ -70,11 +70,24 @@ variable "apply_immediately" {
+variable "availability_zone" {
+  type        = string
+  default     = ""
+  description = "List of Availability Zones for the cluster"
```
Looks like this description is copy / pasta. Mind updating? 

Also, what happens if a module consumer doesn't pass a value for `availability_zone`? Is that acceptable behavior by the provider?

cabrinha replied:
yes, `availability_zone` can just be an empty string. Same with `availability_zones`, an empty list is accpetable.

cabrinha replied:
Updated.

Gowiem reviewed (commented):
One request and one question.

Gowiem commented:
@cabrinha getting the following error from our examples/complete tests: 

```
TestExamplesComplete 2021-04-24T16:40:17Z command.go:121: Error: error creating ElastiCache Cache Cluster: InvalidParameterCombination: Must specify the same number of preferred availability zones as requested number of nodes.
TestExamplesComplete 2021-04-24T16:40:17Z command.go:121: 	status code: 400, request id: 43fe1952-b785-4737-a754-d6c3fdc3b502
TestExamplesComplete 2021-04-24T16:40:17Z command.go:121: 
TestExamplesComplete 2021-04-24T16:40:17Z command.go:121:   on ../../main.tf line 86, in resource "aws_elasticache_cluster" "default":
TestExamplesComplete 2021-04-24T16:40:17Z command.go:121:   86: resource "aws_elasticache_cluster" "default" {
TestExamplesComplete 2021-04-24T16:40:17Z command.go:121: 
TestExamplesComplete 2021-04-24T16:40:17Z command.go:121: 
```

Please look into that and address.

cabrinha commented:
> @cabrinha getting the following error from our examples/complete tests:
> 
> ```
> TestExamplesComplete 2021-04-24T16:40:17Z command.go:121: Error: error creating ElastiCache Cache Cluster: InvalidParameterCombination: Must specify the same number of preferred availability zones as requested number of nodes.
> TestExamplesComplete 2021-04-24T16:40:17Z command.go:121: 	status code: 400, request id: 43fe1952-b785-4737-a754-d6c3fdc3b502
> TestExamplesComplete 2021-04-24T16:40:17Z command.go:121: 
> TestExamplesComplete 2021-04-24T16:40:17Z command.go:121:   on ../../main.tf line 86, in resource "aws_elasticache_cluster" "default":
> TestExamplesComplete 2021-04-24T16:40:17Z command.go:121:   86: resource "aws_elasticache_cluster" "default" {
> TestExamplesComplete 2021-04-24T16:40:17Z command.go:121: 
> TestExamplesComplete 2021-04-24T16:40:17Z command.go:121: 
> ```
> 
> Please look into that and address.

@Gowiem Thanks for the review. I've updated the PR, let's see if the builds go green.

Gowiem commented:
/test all

Gowiem commented:
@cabrinha new issue 😅 

```
TestExamplesComplete 2021-04-26T23:18:20Z command.go:121: 
TestExamplesComplete 2021-04-26T23:18:20Z command.go:121: Error: az_mode "cross-az" is not supported with num_cache_nodes = 1
TestExamplesComplete 2021-04-26T23:18:20Z command.go:121: 
TestExamplesComplete 2021-04-26T23:18:20Z command.go:121:   on ../../main.tf line 86, in resource "aws_elasticache_cluster" "default":
TestExamplesComplete 2021-04-26T23:18:20Z command.go:121:   86: resource "aws_elasticache_cluster" "default" {
TestExamplesComplete 2021-04-26T23:18:20Z command.go:121: 
TestExamplesComplete 2021-04-26T23:18:20Z command.go:121: 
TestExamplesComplete 2021-04-26T23:18:20Z retry.go:80: Returning due to fatal error: FatalError{Underlying: exit status 1}
TestExamplesComplete 2021-04-26T23:18:20Z retry.go:72: terraform [destroy -auto-approve -input=false -lock=false -var-file fixtures.us-east-2.tfvars]
```

nitrocode commented:
Shouldn't the example's `az_mode` be set to `single-az` since that's the original default?

https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/elasticache_cluster#az_mode

cabrinha commented:
/test all

nitrocode commented:
/test all

Gowiem commented:
/test all

Gowiem commented:
@cabrinha 💥 

```
TestExamplesComplete 2021-04-27T22:57:57Z command.go:53: Running command terraform with args [apply -input=false -lock=false -auto-approve -var-file fixtures.us-east-2.tfvars]
TestExamplesComplete 2021-04-27T22:58:00Z command.go:121: 
TestExamplesComplete 2021-04-27T22:58:00Z command.go:121: Error: Reference to undeclared input variable
TestExamplesComplete 2021-04-27T22:58:00Z command.go:121: 
TestExamplesComplete 2021-04-27T22:58:00Z command.go:121:   on main.tf line 16, in module "subnets":
TestExamplesComplete 2021-04-27T22:58:00Z command.go:121:   16:   availability_zones   = var.availability_zones
TestExamplesComplete 2021-04-27T22:58:00Z command.go:121: 
TestExamplesComplete 2021-04-27T22:58:00Z command.go:121: An input variable with the name "availability_zones" has not been declared.
TestExamplesComplete 2021-04-27T22:58:00Z command.go:121: Did you mean "availability_zone"?
TestExamplesComplete 2021-04-27T22:58:00Z command.go:121: 
TestExamplesComplete 2021-04-27T22:58:00Z retry.go:80: Returning due to fatal error: FatalError{Underlying: exit status 1}
TestExamplesComplete 2021-04-27T22:58:00Z retry.go:72: terraform [destroy -auto-approve -input=false -lock=false -var-file fixtures.us-east-2.tfvars]
TestExamplesComplete 2021-04-27T22:58:00Z command.go:53: Running command terraform with args [destroy -auto-approve -input=false -lock=false -var-file fixtures.us-east-2.tfvars]
TestExamplesComplete 2021-04-27T22:58:03Z command.go:121: 
TestExamplesComplete 2021-04-27T22:58:03Z command.go:121: Error: Reference to undeclared input variable
TestExamplesComplete 2021-04-27T22:58:03Z command.go:121: 
TestExamplesComplete 2021-04-27T22:58:03Z command.go:121:   on main.tf line 16, in module "subnets":
TestExamplesComplete 2021-04-27T22:58:03Z command.go:121:   16:   availability_zones   = var.availability_zones
TestExamplesComplete 2021-04-27T22:58:03Z command.go:121: 
TestExamplesComplete 2021-04-27T22:58:03Z command.go:121: An input variable with the name "availability_zones" has not been declared.
TestExamplesComplete 2021-04-27T22:58:03Z command.go:121: Did you mean "availability_zone"?
TestExamplesComplete 2021-04-27T22:58:03Z command.go:121: 
TestExamplesComplete 2021-04-27T22:58:03Z retry.go:80: Returning due to fatal error: FatalError{Underlying: exit status 1}
```

Mind looking at that?

cabrinha commented:
> @cabrinha 💥
> 
> ```
> TestExamplesComplete 2021-04-27T22:57:57Z command.go:53: Running command terraform with args [apply -input=false -lock=false -auto-approve -var-file fixtures.us-east-2.tfvars]
> TestExamplesComplete 2021-04-27T22:58:00Z command.go:121: 
> TestExamplesComplete 2021-04-27T22:58:00Z command.go:121: Error: Reference to undeclared input variable
> TestExamplesComplete 2021-04-27T22:58:00Z command.go:121: 
> TestExamplesComplete 2021-04-27T22:58:00Z command.go:121:   on main.tf line 16, in module "subnets":
> TestExamplesComplete 2021-04-27T22:58:00Z command.go:121:   16:   availability_zones   = var.availability_zones
> TestExamplesComplete 2021-04-27T22:58:00Z command.go:121: 
> TestExamplesComplete 2021-04-27T22:58:00Z command.go:121: An input variable with the name "availability_zones" has not been declared.
> TestExamplesComplete 2021-04-27T22:58:00Z command.go:121: Did you mean "availability_zone"?
> TestExamplesComplete 2021-04-27T22:58:00Z command.go:121: 
> TestExamplesComplete 2021-04-27T22:58:00Z retry.go:80: Returning due to fatal error: FatalError{Underlying: exit status 1}
> TestExamplesComplete 2021-04-27T22:58:00Z retry.go:72: terraform [destroy -auto-approve -input=false -lock=false -var-file fixtures.us-east-2.tfvars]
> TestExamplesComplete 2021-04-27T22:58:00Z command.go:53: Running command terraform with args [destroy -auto-approve -input=false -lock=false -var-file fixtures.us-east-2.tfvars]
> TestExamplesComplete 2021-04-27T22:58:03Z command.go:121: 
> TestExamplesComplete 2021-04-27T22:58:03Z command.go:121: Error: Reference to undeclared input variable
> TestExamplesComplete 2021-04-27T22:58:03Z command.go:121: 
> TestExamplesComplete 2021-04-27T22:58:03Z command.go:121:   on main.tf line 16, in module "subnets":
> TestExamplesComplete 2021-04-27T22:58:03Z command.go:121:   16:   availability_zones   = var.availability_zones
> TestExamplesComplete 2021-04-27T22:58:03Z command.go:121: 
> TestExamplesComplete 2021-04-27T22:58:03Z command.go:121: An input variable with the name "availability_zones" has not been declared.
> TestExamplesComplete 2021-04-27T22:58:03Z command.go:121: Did you mean "availability_zone"?
> TestExamplesComplete 2021-04-27T22:58:03Z command.go:121: 
> TestExamplesComplete 2021-04-27T22:58:03Z retry.go:80: Returning due to fatal error: FatalError{Underlying: exit status 1}
> ```
> 
> Mind looking at that?

Yeah, just pushed another commit. Looks like I renamed the var when I should have just added a new one.

Gowiem commented:
/test all

cabrinha commented:
> /test all

Terratest was expecting two subnets to be returned, but I changed the example to be single-az, so I updated the test to expect a single AZ.

Gowiem commented:
/test all

cabrinha commented:
Looks like we're good here. Anything else I need to do before merge?

Gowiem commented:
Thanks for the contribution @cabrinha! Released as [0.12.0](https://github.com/cloudposse/terraform-aws-elasticache-memcached/releases/tag/0.12.0).
